src/test/java/hudson/plugins/git/extensions/impl/CloneOptionHonorRefSpecTest.java

```diff
@@ -0,0 +1,136 @@
+    public static Collection permuteRefSpecVariable() {
+        List<Object[]> values = new ArrayList<>();
+        /* Should behave same with honor refspec enabled or disabled */
+        boolean honorRefSpec = random.nextBoolean();
+
+        /* Variables set by Jenkins */
+        String[] allowed = {"JOB_BASE_NAME", "JOB_NAME"};
+        for (String refSpecName : allowed) {
+            Object[] combination = {refSpecName, honorRefSpec};
+            values.add(combination);
+            honorRefSpec = !honorRefSpec;
+        }
+
+        /* Variable set by the operating system */
+        String refSpecName = isWindows() ? "USERNAME" : "USER";
+        Object[] combination = {refSpecName, !honorRefSpec};
+        values.add(combination);
+
+        /* Parametrised build */
+        refSpecName = "USER_SELECTED_BRANCH_NAME";
+        combination = new Object[]{refSpecName, !honorRefSpec};
+        values.add(combination);
+
+        return values;
+    }
+
+    /**
+     * This test confirms behavior of refspecs on initial clone with expanded
+     * variables. When an environment variable reference is embedded in the
+     * refspec, it should be expanded in all cases.
+     *
+     * @throws Exception on error
+     */
+    @Test
+    @Issue("JENKINS-56063")
+    public void testRefSpecWithExpandedVariables() throws Exception {
+
+        FreeStyleProject project = createFreeStyleProject();
+        project.addProperty(new ParametersDefinitionProperty(
+                new StringParameterDefinition("USER_SELECTED_BRANCH_NAME", "user_branch")
+        ));
+        project.save();
+
+        String refSpecExpectedValue = getExpectedValue(refSpecName, project);
+
+        // create initial commit
+        final String commitFile1 = "commitFile1";
+        commit(commitFile1, johnDoe, "Commit in master branch");
+
+        // create branch and make initial commit
+        git.checkout().ref("master").branch(refSpecExpectedValue).execute();
+        commit(commitFile1, johnDoe, "Commit in '" + refSpecExpectedValue + "' branch");
+
+        // Use the variable reference in the refspec
+        // Should be expanded by the clone option whether or not honor refspec is enabled
+        List<UserRemoteConfig> repos = new ArrayList<>();
+        repos.add(new UserRemoteConfig(
+                testRepo.gitDir.getAbsolutePath(),
+                "origin",
+                "+refs/heads/${" + refSpecName + "}:refs/remotes/origin/${" + refSpecName + "}", null));
+
+        /* Use the variable or its value as the branch name.
+         * Same result expected in either case.
+         */
+        String branchName = random.nextBoolean() ? "${" + refSpecName + "}" : refSpecExpectedValue;
+        GitSCM scm = new GitSCM(
+                repos,
+                Collections.singletonList(new BranchSpec(branchName)),
+                false, Collections.<SubmoduleConfig>emptyList(),
+                null, JGitTool.MAGIC_EXENAME,
```

Comment:
Very nice that you included a check of the JGit implementation!  Since we expect JGit and command line git to work interchangeably in this case, could we choose the git implementation randomly for each test?  That increases the chances that both CLI git and JGit will be evaluated in at least one of the test cases.
```suggestion
                null, random.nextBoolean() ? JGitTool.MAGIC_EXENAME : null,
```

Reply:
(Brought that code outside of method `setupProject` so that we could get the build name without any issues, with an explicit freestyle project)

Will do, thanks!